### PR TITLE
New class Url as addition for a later Puli PR (2nd attempt)

### PR DESCRIPTION
**Pull request by SenseException**
Changes suggested in #12 are now commited.

Should `Url` be a part of the current readme?

### CODE REVIEW

File: src/Url.php [hunk 1]
```diff
@@ -99,8 +95,10 @@ private static function split($url)
                 $url = '';
             }
         } else {
-            $scheme = '';
-            $domain = '';
+            throw new InvalidArgumentException(sprintf(
```
Comment (webmozart) [inline src/Url.php]: Could you invert the if condition and move this to the if-block? Then you can get rid of the else-block.

**Comment by webmozart** [comment]
Thanks :) Yes, adding it to the README would be excellent. Also, please rebase this onto master.

**Comment by SenseException** [comment]
@webmozart New Update for the PR.

**Comment by SenseException** [comment]
Push.

File: src/Url.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+     * @param string $baseUrl A base path or URL.
+     *
+     * @return string
+     *
+     * @throws InvalidArgumentException If the hostname of url and baseUrl does
+     *                                  not match.
+     */
+    public static function makeRelative($url, $baseUrl)
+    {
+        Assert::string($url, 'The path must be a string. Got: %s');
```
Comment (webmozart) [inline src/Url.php]: "The URL", "The base URL"

File: src/Url.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                '"%s" is not a valid Url.',
+                $url
+            ));
+        }
+
+        $scheme = substr($url, 0, $pos + 3);
+        $url = substr($url, $pos + 3);
+
+        if (false !== ($pos = strpos($url, '/'))) {
+            $domain = substr($url, 0, $pos);
```
Comment (webmozart) [inline src/Url.php]: I'd rename `$domain` to `$host` which is the common terminology when speaking about URLs.

File: src/Url.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                $root,
+                $baseRoot
+            ));
+        }
+
+        return $path;
+    }
+
+    /**
+     * Splits a part into its root directory and the remainder.
```
Comment (webmozart) [inline src/Url.php]: Splits a URL into its host and the path.

File: src/Url.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+     *
+     * If the path has no URL, an empty root directory will be
+     * returned.
```
Comment (webmozart) [inline src/Url.php]: This does not seem to be accurate.

File: src/Url.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+     *
+     * list ($root, $path) = Path::split("http://example.com/webmozart")
```
Comment (webmozart) [inline src/Url.php]: Can you add

    ```

before this line and after the code to enable proper code output in the rendered API?

File: src/Url.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+     * // => array("http://example.com", "/webmozart")
+     *
+     * list ($root, $path) = Path::split("http://example.com")
+     * // => array("http://example.com", "")
+     *
+     * @param string $url The URL to split.
+     *
+     * @return string[] An array with the domain and the remaining
```
Comment (webmozart) [inline src/Url.php]: with the host and the path of the URL

File: src/Url.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+     */
+    public static function makeRelative($url, $baseUrl)
+    {
+        Assert::string($url, 'The path must be a string. Got: %s');
+        Assert::string($baseUrl, 'The base path must be a string. Got: %s');
+
+        list($root, $relativePath) = self::split($url);
+        list($baseRoot, $relativeBasePath) = self::split($baseUrl);
+
+        $path = Path::makeRelative($relativePath, $relativeBasePath);
```
Comment (webmozart) [inline src/Url.php]: You can move this line below the `if` block.

File: src/Url.php [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+        if ('' !== $root && '' !== $baseRoot && $root !== $baseRoot) {
+            throw new InvalidArgumentException(sprintf(
+                'Domain "%s" doesn\'t equal to base "%s".',
```
Comment (webmozart) [inline src/Url.php]: The URL "%s" cannot be made relative to "%s" since their host names are different.

File: src/Url.php [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+     * @throws InvalidArgumentException If the hostname of url and baseUrl does
+     *                                  not match.
+     */
+    public static function makeRelative($url, $baseUrl)
+    {
+        Assert::string($url, 'The path must be a string. Got: %s');
+        Assert::string($baseUrl, 'The base path must be a string. Got: %s');
+
+        list($root, $relativePath) = self::split($url);
+        list($baseRoot, $relativeBasePath) = self::split($baseUrl);
```
Comment (webmozart) [inline src/Url.php]: `$root` => `$host`, `$baseRoot` => `$baseHost`, `$relativePath` => `$path`, `$relativeBasePath` => `$basePath`

**Comment by webmozart** [comment]
Looks good, thanks @SenseException! :) Can you integrate my comments?

**Comment by SenseException** [comment]
Done, but I get a StyleCI Error: "The StyleCI analysis was misconfigured".

**Comment by webmozart** [comment]
I think you need to rebase your branch on master.

**Comment by SenseException** [comment]
StyleCI is green again.

File: tests/UrlTest.php [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        $host = 'ftp://user:password@example.com:8080';
+
+        $relative = Url::makeRelative($host.$absolutePath, $host.$basePath);
+        $this->assertSame($relativePath, $relative);
+    }
+
+    /**
+     * @expectedException \InvalidArgumentException
+     * @expectedExceptionMessage "/webmozart/puli/css/style.css" is not a valid Url.
```
Comment (webmozart) [inline tests/UrlTest.php]: This message is a bit misleading. I'd expect "is not an absolute URL".

File: tests/UrlTest.php [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+     * @covers Webmozart\PathUtil\Url
+     */
+    public function testMakeRelativeWithoutFullUrl()
+    {
+        $relative = Url::makeRelative('/webmozart/puli/css/style.css', 'http://example.com/webmozart/puli');
```
Comment (webmozart) [inline tests/UrlTest.php]: Unused variable.

Comment (webmozart) [inline tests/UrlTest.php]: What do you think about assuming that the path is on the same host if no host name is included? i.e.

```php
Url::makeRelative('/webmozart/puli/css/style.css', 'http://example.com/webmozart/puli');
// => css/style.css
```

For relative paths, we could simply return the relative path unchanged:

```php
Url::makeRelative('css/style.css', 'http://example.com/webmozart/puli');
// => css/style.css
```

Comment (SenseException) [inline tests/UrlTest.php]: I don't see any disadvantages by that. The unchanged relative path would be consistent to `Path::makeRelative`. Do you have a specific use case in mind suggesting this behaviour?

Comment (webmozart) [inline tests/UrlTest.php]: Mostly the case where the call to `Url::makeRelative()` is hidden behind some generic function, like `url()` in a template. There a user could expect to be able to pass both URLs and relative/absolute paths (expecting URLs for the current host).

Comment (SenseException) [inline tests/UrlTest.php]: @webmozart Could you please take a look on the new changes?

I decided to not simply return a relative path because of paths like `css/../style.css` should be normalized and I don't want to use another method from `Path` in `Url::makeRelative` to prevent another static dependency. `Path::makeRelative` is handling this internally.

File: tests/UrlTest.php [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        );
+    }
+
+    public function provideMakeRelativeIsAlreadyRelativeTests()
+    {
+        return array(
+            array('css/style.css', '/webmozart/puli', 'css/style.css'),
+            array('css/style.css', '', 'css/style.css'),
+            array('css/../style.css', '', 'style.css'),
+            array('css/./style.css', '', 'css/style.css'),
```
Comment (webmozart) [inline tests/UrlTest.php]: I think you should also cover the edge cases:

```php
array('../style.css', '', 'style.css'),
array('./style.css', '', 'style.css'),
```

**Comment by webmozart** [comment]
Apart from my last comment, this looks good now! :) Can you think of any more edge cases?

**Comment by SenseException** [comment]
I'll take a look which further paths I can cover in the tests at the end of the week. When this first step is done, I'll take on the Puli PR that is going to use `Url`.

**Comment by SenseException** [comment]
@webmozart The following provided data fails in the test.

``` php
array('../style.css', '', 'style.css'),
```

but has also the same behaviour in `Path` class with

``` php
array('../style.css', '/', 'style.css'),
```

In both classes makeRelative returns `.../style.css` instead. Is this result expected from `Path` and `Url` is supposed to act differently than `Path`, or is this a bug in `Path::makeRelative`?

If this is a bug in `Path::makeRelative`, I would prefer to handle this in another PR, since this PR is supposed to handle `Url` and both makeRelative-methods should act the same for user experience.

**Comment by webmozart** [comment]
I think the expected data should be:

```php
array('../style.css', '', '../style.css'),
array('../style.css', '/', 'style.css'),
```

The `Path` classes treat empty strings like the "current directory" (`.` on Unix), so the first case should return the relative path unchanged. The second path has a clear base - the root. Since we can't move outside the root, the dot segments should be stripped.

Does this make sense?

**Comment by SenseException** [comment]
``` php
array('../style.css', '/', 'style.css'),
```
This is where `Path::makeRelative` fails. It returns `../style.css` instead. That's why also `Url::makeRelative` will fail because it uses Path. I would prefer to create a new PR to fix `Path::makeRelative` instead of fixing it in this PR,  since this PR is about adding `Url::makeRelative`.

**Comment by webmozart** [comment]
I see. Should we fix this in a separate PR and then rebase this one once merged?

**Comment by SenseException** [comment]
Sounds great.

**Comment by webmozart** [comment]
Cool :) Can you work on this PR?

**Comment by SenseException** [comment]
That was my plan :-)

**Comment by SenseException** [comment]
I rebased the current master and put some more provider data into the tests.

**Comment by webmozart** [comment]
Great, thanks @SenseException! :) I think this is ready to merge now.

**Comment by SenseException** [comment]
You've earned a "Yay" badge \o/
